35-CAN-vehicle-monitor/firmware/src/main.cpp

```diff
@@ -0,0 +1,206 @@
+
+// Uncomment this line and replace com.your-company:your-product-name with your
+// ProductUID.
+// #define PRODUCT_UID "com.my-company.my-name:my-project"
+
+#ifndef PRODUCT_UID
+#define PRODUCT_UID ""
+#pragma message "PRODUCT_UID is not defined in this example. Please ensure your Notecard has a product identifier set before running this example or define it in code here. More details at https://bit.ly/product-uid"
+#endif
+
+#define ENV_VAR_POLL_MS (15 * 1000) // 15 seconds
+
+#define REQUESTID_TEMPLATE 1
+#define APPLICATION_NOTEFILE "data.qo"
+
+// Extended IDs are 29 bits. Rounding that up to 32 bits yields 4 bytes.
+// 2: 0x prefix.
+// MAX_CAN_ID_BYTES * 2: 2 hex characters for each byte.
+#define MAX_CAN_ID_CHARS 10
+
+#define MAX_CAN_DATA_BYTES 8
+// MAX_CAN_DATA_BYTES * 2: 2 hex characters for each byte.
+// 7: colon delimiters between bytes.
+#define MAX_CAN_DATA_CHARS 23
+
+Notecard notecard;
+CanIdManager canIdManager = CanIdManager(notecard, ENV_VAR_POLL_MS);
+
+// Register the notefile template for our data.
+static bool registerNotefileTemplate()
+{
+    notecard.logDebugf("Registering %s template.\r\n", APPLICATION_NOTEFILE);
+    J *req = notecard.newRequest("note.template");
+    if (req == NULL) {
+        notecard.logDebug("Failed to create note.template request.\r\n");
+        return false;
+    }
+
+    J *body = JCreateObject();
+    if (body == NULL) {
+        JDelete(req);
+        notecard.logDebug("Failed to create note.template request body.\r\n");
+        return false;
+    }
+
+    // Add an ID to the request, which will be echo'ed
+    // back in the response by the notecard itself.  This
+    // helps us to identify the asynchronous response
+    // without needing to have an additional state.
+    JAddNumberToObject(req, "id", REQUESTID_TEMPLATE);
+    JAddStringToObject(req, "file", APPLICATION_NOTEFILE);
+    JAddStringToObject(body, "id", TSTRING(MAX_CAN_ID_CHARS));
+    JAddStringToObject(body, "data", TSTRING(MAX_CAN_DATA_CHARS));
+    JAddItemToObject(req, "body", body);
+
+    if (!notecard.sendRequest(req)) {
+        notecard.logDebug("Failed to send note.template request.\r\n");
+        return false;
+    }
+
+    notecard.logDebug("Template registration succeeded.\r\n");
+
+    return true;
+}
+
+static bool sendDateNote(uint32_t id, const uint8_t* data, int len)
+{
+    J *req = notecard.newRequest("note.add");
+    if (req == NULL) {
+        notecard.logDebug("Failed to create note.add request.\r\n");
+        return false;
+    }
+
+    J *body = JCreateObject();
+    if (body == NULL) {
+        JDelete(req);
+        notecard.logDebug("Failed to create note.add request body.\r\n");
+        return false;
+    }
+
+    JAddStringToObject(req, "file", APPLICATION_NOTEFILE);
+
+    char idStr[MAX_CAN_ID_CHARS + 1]; // +1 for \0.
+    sprintf(idStr, "0x%08lx", id);
+    JAddStringToObject(body, "id", idStr);
+
+    char dataStr[MAX_CAN_DATA_CHARS + 1]; // +1 for \0.
```

Comment:
Would it make sense to represent the payload data as an array rather than a string? I imagine it depends upon how it's consumed upstream, but certainly for jsonata it would make the data easier to consume.